### ISSUE TITLE
Cannot paste content copied from outside on VS Code Web

### ISSUE THREAD

**Issue by guiherzog** [microsoft/vscode#200641]
- VS Code Version: Version: 1.85.0
Commit: af28b32d7e553898b2a91af498b1fb666fdebe0c
Date: 2023-12-06T18:18:26.691Z (5 days ago)
- OS Version:  Browser: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36

Steps to Reproduce:

1. Open vscode.dev (or insiders).
2. Copy a file with cmd/ctrl+c in the file explorer.
3. Paste inside a document.
4. It should paste the file path.
5. Switch to another browser tab that is not vscode.dev and copy any text inside of it with cmd/ctrl+c.
6. Switch back to vscode.dev.
7. Paste the copied content inside an opened file in the editor.

Expected result:
The text copied outside of VS Code should be pasted there.

Current result:
It pastes the path of the file copied by step 2 in the file explorer.

This blocks me from using copy & paste properly until I reload VS Code completely.

**Comment by lramos15** [comment]
@mjbvz Seems like we're not properly reading the right contents when pasting in the editor

1. Copy a file 
2. Paste into notepad 
3. Nothing
---------------------
1. Copy a file
2. Paste into editor, get paste options for which path type
---------------------
1. Copy a file
2. Copy other text from a different source
3. Paste into editor.
4. Still the file path
---------------------
1. Copy a file
2. Copy other text from a different source
3. Paste into other application.
4. The text in the clipboard is properly pasted

**Comment by mjbvz** [comment]
Seems caused by the web clipboard being in-memory: 

https://github.com/microsoft/vscode/blob/fd802370822b1bbd8d5eed899cdc8a50a63da436/src/vs/platform/clipboard/browser/clipboardService.ts#L152

In the copy paste controller, I explicitly clear it to work around this:

https://github.com/microsoft/vscode/blob/348562692b1e6d2ccf026e7c7a84e88616e799d8/src/vs/editor/contrib/dropOrPasteInto/browser/copyPasteController.ts#L127

However that won't work for external copies

**Comment by bpasero** [comment]
This is tough to get right using web APIs where support for custom formats is very limited. Our solution is not very robust and thus leads to this issue.

I have played around for 10 minutes writing a different type into the clipboard for the `writeResources` / `readResources` method:

```ts
async writeResources(resources: URI[]): Promise<void> {
		try {
			return await navigator.clipboard.write([
				new ClipboardItem({
					['text/html']: new Blob([`<vscode>${resourcesToBuffer(resources).toString()}</vscode>`], { type: 'text/html' }),
				})
			]);
		} catch (error) {
			console.error(error);
		}

		this.resources = resources;
	}

	async readResources(): Promise<URI[]> {
		try {
			const clipboardItems = await navigator.clipboard.read();
			for (const clipboardItem of clipboardItems) {
				for (const type of clipboardItem.types) {
					if (type !== 'text/html') {
						continue;
					}

					const blob = await clipboardItem.getType(type);
					console.log(await blob.text());
				}
			}
		} catch (error) {
			console.error(error);
		}
		return this.resources;
	}
```

One immediate issue is that `text/uri-list` for example is not a valid data type in Chrome. You can only use `text/plain`, `text/html` and `image/png` (according to https://stackoverflow.com/questions/68897154/list-of-supported-mime-types-for-clipboard-write).

We could maybe encode this into `text/html` and then try to decode it with a special format to see if it matches our format to implement this, but I am not sure if its worth it.

Maybe @deepak1556 for additional thoughts.

**Comment by bpasero** [comment]
I think this is actually quite a severe issue in web. Once you have resources in your clipboard, it seems to me you cannot paste anything else into the editor but that (!). So for example, these steps are much easier to reproduce:
* copy in the explorer
* decide to "Copy Path" from the explorer context menu
* paste into editor
* 🐛  the path is not pasted

One approach to solve this (see https://github.com/microsoft/vscode/pull/200782) is to associate the resources with a state that is specific to the contents of the clipboard at that time with the goal to detect if that state has changed or not. Its ugly, because then `writeResources` already needs access to the clipboard, but I think the current behaviour is worse.

https://github.com/microsoft/vscode/blob/65086954107f34a3e253b3dcfe9ec217490eb98b/src/vs/platform/clipboard/browser/clipboardService.ts#L181-L192

**Comment by bpasero** [comment]
I am also adding code to reset the `resources` when any `copy` event is detected in the document:

https://github.com/microsoft/vscode/blob/33f67c27747845f418235a6d77c14d05ba8e0879/src/vs/platform/clipboard/browser/clipboardService.ts#L32-L38

I believe that would make the code here obsolete:

https://github.com/microsoft/vscode/blob/33f67c27747845f418235a6d77c14d05ba8e0879/src/vs/editor/contrib/dropOrPasteInto/browser/copyPasteController.ts#L130

**Comment by guiherzog** [comment]
Thank you for the quick resolution!

**Comment by bpasero** [comment]
This should work fine now from our insiders build if you want to verify: https://insiders.vscode.dev/

**Comment by guiherzog** [comment]
It works for me, thanks!

**Comment by bilogic** [comment]
Was this released into 1.85.1?
